src/VariablesState.ts

```diff
@@ -0,0 +1,311 @@
+			callback(variableName, newValue);
+		});
+	}
+
+	get batchObservingVariableChanges(){
+		return this._batchObservingVariableChanges;
+	}
+	set batchObservingVariableChanges(value: boolean){
+		value = !!value; // TODO: Might have become useless with TypeScript.
+		this._batchObservingVariableChanges = value;
+		if (value) {
+			this._changedVariables = new Set();
+		}
+
+		else {
+			if (this._changedVariables != null) {
+				for (let variableName of this._changedVariables) {
+					let currentValue = this._globalVariables.get(variableName);
+					if (!currentValue) {
+						throwNullException('currentValue');
+					} else {
+						this.variableChangedEvent(variableName, currentValue);
+					}
+				}
+			}
+		}
+	}
+
+	get callStack(){
+		return this._callStack;
+	}
+	set callStack(callStack: any /* CallStack */){
+		this._callStack = callStack;
+	}
+
+	private _batchObservingVariableChanges: boolean = false;
+
+	// the original code uses a magic getter and setter for global variables,
+	// allowing things like variableState['varname]. This is not quite possible
+	// in js without a Proxy, so it is replaced with this $ function.
+	public $(variableName: string, value: InkObject){
+		if (typeof value === 'undefined'){
+			let varContents = this._globalVariables.get(variableName);
+
+			if ( typeof varContents === 'undefined' ) {
+				varContents = this._defaultGlobalVariables.get(variableName);
+			}
+
+			if ( typeof varContents !== 'undefined' )
+				return (varContents as AbstractValue).valueObject;
+			else
+				return null;
+		}
+		else{
+			if (typeof this._defaultGlobalVariables.get(variableName) === 'undefined')
+				throw new StoryException("Cannot assign to a variable that hasn't been declared in the story");
+
+			let val = Value.Create(value);
+			if (val == null) {
+				if (value == null) {
+					throw new StoryException('Cannot pass null to VariableState');
+				} else {
+					throw new StoryException('Invalid value passed to VariableState: '+value.toString());
+				}
+			}
+
+			this.SetGlobal(variableName, val);
+		}
+	}
+
+	constructor(callStack: any /* CallStack */, listDefsOrigin: ListDefinitionsOrigin){
+		this._globalVariables = new Map();
+		this._callStack = callStack;
+		this._listDefsOrigin = listDefsOrigin;
+
+		// TODO should we discourage the use of proxies since they're non-portable
+		// and direct access is unsafe? if es6 proxies are available, use them.
+		try{
+			// the proxy is used to allow direct manipulation of global variables.
+			// It first tries to access the objects own property, and if none is
+			// found it delegates the call to the $ method, defined below
+			let p = new Proxy(this, {
+				get(target: any, name){
+					return (name in target) ? target[name] : target.$(name);
+				},
+				set(target: any, name, value){
+					if (name in target) target[name] = value;
+					else target.$(name, value);
+					return true; // returning a falsy value make the trap fail
```

Comment:
I wonder what I meant by that, but it doesn't look very sound.

Reply:
Maybe we should rework the entire proxy thing later?

Reply:
Yah, we'll see :)